barcodes_search/wizards/barcode_search.py

```diff
@@ -0,0 +1,121 @@
+        that matches with the given barcode
+        """
+        barcode_fields = self.get_barcode_fields()
+        res = self._search_by_barcode_barcode_fields(
+            barcode, barcode_fields)
+
+        # Extra search for objects that have a barcode with encoded data
+        # Typical exemple, product with a pattern 20{barcode_base}{NNDDD}
+        # where {NNDDD} is the weight of the product
+        nomenclature_obj = self.env['barcode.nomenclature']
+        nomenclatures = nomenclature_obj.search([])
+        rule_types = self.get_model_by_rule_type()
+        for nomenclature in nomenclatures:
+            parsed_result = nomenclature.parse_barcode(barcode)
+            rule_type = parsed_result['type']
+            if rule_type in rule_types:
+                barcode_fields = self.get_barcode_fields(rule_types[rule_type])
+                base_code = parsed_result['base_code']
+                extra_results = self._search_by_barcode_barcode_fields(
+                    base_code, barcode_fields)
+                for extra_result in extra_results:
+                    extra_result[2]['type'] = rule_type
+                    extra_result[2]['value'] = parsed_result['value']
+                res += extra_results
+        return res
+
+    @api.model
+    def get_barcode_fields(self, model_name=False):
+        """Return a recordset of fields that represent a barcode, in any model.
+        By default, it will return all fields named 'barcode'.
+        Note : Overload that function in a custom module, if you define
+        a barcode field with a name different than 'barcode'"""
+        field_obj = self.env['ir.model.fields']
+        if model_name:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '=', model_name)]
+        else:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '!=', 'barcode.search')]
+        return field_obj.search(domain)
+
+    @api.model
+    def get_model_by_rule_type(self):
+        """Return a dictionary of rule type: model
+        for each special barcode nomenclature rule types, with  a pattern
+        that belong a fixed part like '.....' and a variable part like {NNDDD}.
+        Note : Overload that function in a custom module, if you define
+        such new rule type.
+
+        'weight' is define in Odoo / 'stock' module.
+        'price' is define in Odoo / 'point_of_sale' module.
+        'price_to_weight' is defined in OCA / 'pos_price_to weight' module
+        """
+        return {
+            'weight': 'product.product',
+            'price': 'product.product',
+            'price_to_weight': 'product.product',
+        }
+
+    # Private Section
+    @api.model
+    def _search_by_barcode_barcode_fields(self, barcode, barcode_fields):
+        res = []
+        for barcode_field in barcode_fields:
+            model_obj = self.env.get(barcode_field.model, False)
+            if not type(model_obj) is bool:
```

Comment:
This is not a good way to test typing in Python. Typically you want to use `isinstance(` in order to allow for inheritance. Was there a reason to not use that here?

Reply:
Ahh looking at this more, you're using this to determine if there was a `barcode_field.model` in the env? If so, typically EAFP is considered the more Pythonic approach:

```python
try:
    model_obj = self.env[barcode_field.model]
except KeyError:
    _logger.debug(
        'The %s model was not found in the env.', barcode_field.model,
    )
    continue
items = model_obj.search([(barcode_field.name, '=', barcode)])
res += [
    (barcode_field, item, {}) for item in items
]
```
  
Note the logger, which is to follow "Errors should never pass silently, unless explicitly silenced" of [PEP-20](https://www.python.org/dev/peps/pep-0020/). IMO the model not existing is an error, but I could be misunderstanding that part.

Reply:
It's not an error. It can occure "normaly" especially in tests cases, if you have other module installed because when the test is executed, only a part of model are loaded in env, but all are in database. So I don't think a KeyError is necessary.

Reply:
Thanks for the explanation @legalsylvain. Just the `_logger` was the part indicating the error though, so that would just be omitted if it's a normal case. The `KeyError` is just a Python idiom for [EAFP](https://docs.python.org/2/glossary.html#term-eafp). Assuming that a key exists and reacting if it doesn't is way cheaper in computational complexity than checking for it to exist.